File: LeetCode/DP/Golang/Leetcode730-count-different-palindromic-subsequences.go
package sbgo

func countPalindromicSubsequences(s string) int {
	n, P := len(s), int(1e9)+7
	f := make([][]int, n+1)
	for i := 0; i <= n; i++ {
		f[i] = make([]int, n+1)
		for j := 0; j <= n; j++ {
			f[i][j] = 1
			if i == j {
				f[i][j]++
			}
		}
	}
	for length := 2; length <= n; length++ {
		q := make([][]int, 4)
		for j := 1; j <= n; j++ {
			q[s[j-1]-'a'] = append(q[s[j-1]-'a'], j)
			i := j - length + 1
			if i >= 1 {
				for k := 0; k < 4; k++ {
					for len(q[k]) > 0 && q[k][0] < i {
						q[k] = q[k][1:]
					}
					if len(q[k]) > 0 {
						f[i][j]++
						l, r := q[k][0], q[k][len(q[k])-1]
						if l < r {
							f[i][j] = (f[i][j] + f[l+1][r-1]) % P
						}
					}
				}
			}
		}
	}
	return (f[1][n] - 1 + P) % P
}

File: LeetCode/DP/JS/Leetcode730-count-different-palindromic-subsequences.js
var countPalindromicSubsequences = function (s) {
    const n = s.length, P = parseInt(1e9) + 7;
    let f = new Array(n + 1).fill(0).map(() => new Array(n + 1).fill(1));
    for (let i = 1; i <= n; i++) f[i][i] += 1;
    for (let len = 2; len <= n; len++) {
        let q = new Array(4).fill(0).map(() => new Array());
        for (let j = 1; j <= n; j++) {
            q[s[j - 1].charCodeAt() - 'a'.charCodeAt()].push(j);
            let i = j - len + 1;
            if (i >= 1) {
                for (let k = 0; k < 4; k++) {
                    while (q[k].length && q[k][0] < i) q[k].shift();
                    if (q[k].length) {
                        f[i][j]++;
                        let [l, r] = [q[k][0], q[k][q[k].length - 1]];
                        if (l < r) {
                            f[i][j] = (f[i][j] + f[l + 1][r - 1]) % P;
                        }
                    }
                }
            }
        }
    }
    return (f[1][n] - 1 + P) % P;
};